nesta/packages/nomis/nomis.py

```diff
@@ -0,0 +1,249 @@
+
+def get_base(x):
+    """Get the NOMIS column name base entity, for columns named '{something}_code'.
+    
+    Args:
+        x (str): The NOMIS column name, of the form {something}_code
+    Returns:
+        base (str): The NOMIS column name base entity
+    """
+    return REGEX.findall(x)[0]
+
+
+def get_code_pairs(columns):
+    """Pair NOMIS key-value column names.
+
+    Args:
+        columns (list): A list of NOMIS column names.
+    Returns:
+        cols (dict): Mapping of key-value column names.
+    """
+    cols = {c: f"{get_base(c)}_name"
+            for c in columns if c.endswith("_code")
+            if f"{get_base(c)}_name" in columns}
+    return cols
+
+
+def reformat_nomis_columns(data):
+    """Reformat columns from default NOMIS data.
+
+    Args:
+        df (:obj:`pd.DataFrame`): Dataframe containing NOMIS data.
+    Returns:
+        tables (:obj:`list` of :obj:`dict`): Reformatted rows of data.
+    """
+    tables = defaultdict(list)
+    # Generate the NOMIS key-value column pairs, and recursively
+    # replace them
+    for name, df in data.items():
+        _df = df.copy()  # Don't mess with the original data
+        pairs = get_code_pairs(_df.columns)
+        for _code, _name in pairs.items():
+            base = f"{get_base(_code)}_lookup"
+            df_codes = _df[[_name,_code]].copy()
+            df_codes.columns = ["name","code"]
+            tables[base] += df_codes.to_dict(orient="records")
+            _df.drop(_name, axis=1, inplace=True)
+        tables[name] = _df.to_dict(orient="records")
+        for row in tables[name]:
+            row['date'] = row['date'].to_pydatetime()
+    # Append this pair of values
+    return tables
+
+def batch_request(config, dataset_id, geographies, date_format, 
+                  record_offset=0, max_api_calls=10):
+    """Fetch a NOMIS dataset from the API, in batches, 
+    based on a configuration object.
+    
+    Args:                                                               
+        config (dict): Configuration object, from which a get
+                       request is formed.
+        dataset_id (str): NOMIS dataset ID
+        geographies (list): Return object from :obj:`discovery_iter`.
+        date_format (str): Formatting string for dates in the dataset
+    Returns:                                                            
+        dfs (:obj:`list` of :obj:`pd.DataFrame`): Batch return results.      
+    """
+    config["geography"] = ",".join(str(row["nomis_id"]) 
+                                   for row in geographies)
+    config["RecordOffset"] = record_offset
+    date_parser = lambda x: pd.datetime.strptime(x, date_format)
+
+    # Build a list of dfs in chunks from the NOMIS API
+    dfs = []
+    offset = 25000
+    icalls = 0
+    while offset > 0 and icalls < max_api_calls:
+        #logging.debug(f"\t\t {offset}")
+        # Build the request payload
+        params = "&".join(f"{k}={v}" for k,v in config.items())
```

Comment:
Normally yes, but in this case there was some strange URL encoding issues with the config file (I think because there are commas in the geography parameter)